plugins/flutter_plugins/url_launcher_fde/gtk/url_launcher_plugin.cc

```diff
@@ -0,0 +1,152 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "include/url_launcher/url_launcher_plugin.h"
+
+#include <flutter_linux/flutter_linux.h>
+#include <gtk/gtk.h>
+
+// See url_launcher_channel.dart for documentation.
+const char kChannelName[] = "plugins.flutter.io/url_launcher";
+const char kBadArgumentsError[] = "Bad Arguments";
+const char kLaunchError[] = "Launch Error";
+const char kCanLaunchMethod[] = "canLaunch";
+const char kLaunchMethod[] = "launch";
+const char kUrlKey[] = "url";
+
+struct _FlUrlLauncherPlugin {
+  GObject parent_instance;
+
+  FlPluginRegistrar* registrar;
+
+  // Connection to Flutter engine.
+  FlMethodChannel* channel;
+};
+
+G_DEFINE_TYPE(FlUrlLauncherPlugin, fl_url_launcher_plugin, g_object_get_type())
+
+// Gets the URL from the arguments or generates an error.
+static gchar* get_url(FlValue* args, GError** error) {
+  if (fl_value_get_type(args) != FL_VALUE_TYPE_MAP) {
+    g_set_error(error, 0, 0, "Argument map missing or malformed");
+    return nullptr;
+  }
+  FlValue* url_value = fl_value_lookup_string(args, kUrlKey);
+  if (url_value == nullptr) {
+    g_set_error(error, 0, 0, "Missing URL");
+    return nullptr;
+  }
+
+  return g_strdup(fl_value_get_string(url_value));
+}
+
+// Called to check if a URL can be launched.
+static FlMethodResponse* can_launch(FlUrlLauncherPlugin* self, FlValue* args) {
+  g_autoptr(GError) error = nullptr;
+  g_autofree gchar* url = get_url(args, &error);
+  if (url == nullptr) {
+    return FL_METHOD_RESPONSE(fl_method_error_response_new(
+        kBadArgumentsError, error->message, nullptr));
+  }
+
+  gboolean is_launchable = FALSE;
+  g_autofree gchar* scheme = g_uri_parse_scheme(url);
+  if (scheme != nullptr) {
```

Comment:
If there's no scheme, it wasn't a valid URI, so failure of this condition should be respond with an error rather than `success(false)`.